### PR TITLE
Refactor arc and rect to use canvas methods rather than curves

### LINKED ISSUE #6485: Update and simplify implementation of arc, ellipse, and possibly rect
### Increasing Access

This enhancement would simplify p5's source code and would reduce the mathematical prerequisites for understanding it. This would increase access among p5 contributors.

### Most appropriate sub-area of p5.js?

- [ ] Accessibility
- [ ] Color
- [X] Core/Environment/Rendering
- [ ] Data
- [ ] DOM
- [ ] Events
- [ ] Image
- [ ] IO
- [ ] Math
- [ ] Typography
- [ ] Utilities
- [ ] WebGL
- [ ] Build Process
- [ ] Unit Testing
- [ ] Internalization
- [ ] Friendly Errors
- [ ] Other (specify if possible)

### Feature enhancement details

**Proposal**
Use modern features of the native canvas API to simplify the implementation of `arc`, `ellipse`, and possibly `rect`.

**The problem**
The 2D renderer's current implementation of `arc` and `ellipse` is complicated by the fact that it uses Bézier curves, which cannot exactly represent either primitive. The implementation is complicated enough that the [source comments](https://github.com/processing/p5.js/blob/main/src/core/p5.Renderer2D.js#L507) link to an outside paper for an explanation of the underlying mathematics.

The implementation of `rect` may also be more complicated than necessary, since rounded corners require clipping the radii and building the rectangle with the canvas's `arcTo` method.

**The solution**
The use of Bézier curves in p5's 2D implementation of `arc` and `ellipse` dates to [p5's addition of `ellipse`](https://github.com/processing/p5.js/commit/213d56dbaabee03f3ebd44ec358b0c0fa2cb1fef) on July 4, 2013. At that time, the canvas didn't have a native ellipse method. Now it does. The canvas's [`ellipse`](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/ellipse) method actually draws elliptical arcs, so it could be used to implement both `arc` and `ellipse` in p5. This would be a significant simplification. Also, `ellipse` has been available in Chrome [since November 2013](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D#browser_compatibility), so bugs and browser compatibility don't seem like a significant concern.

The situation with `rect` is similar. After p5 already supported rectangles with rounded corners, [the `roundRect` method](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/roundRect) was added to the native Canvas API. However, this addition is much more recent (browsers began supporting it in 2022, with Firefox support only being added in 2023).

**Update:**
Here's a task list to help us keep track of what work has been completed, since @tuminzee has already done the work for `ellipse()` (it just needs to be merged). Note that the task list does include `rect()`, based on the comments on this issue.

- [x] [Update and simplify implementation of `ellipse()`](https://github.com/processing/p5.js/pull/6499)
- [ ] Update and simplify implementation of `arc()`
- [ ] Update and simplify implementation of `rect()`

### PR DESCRIPTION
**Pull request by ksen0**
Resolves #6485 

Changes:

* in Renderer2D `arc`, use canvas `ellipse`, removing the need for `_acuteArcToBezier`
* in Renderer2D `rect`, use canvas `roundRect` when there are rounded corners, rater than `ArcTo`

 Screenshots of the change:

image

SCREENSHOT: Same code running with current p5 (left) and locally (right) demonstrating visual equivalence for `rect` with different rounded corners, one of them clipped.

#### PR Checklist

- [x] `npm run lint` passes
- [x] [Inline documentation] is included / updated
- [ ] [Unit tests] are included / updated

No additional unit tests have been included*. Manual testing was done with all the standard examples from the p5 documentation on both `arc` and `rect` as well as some additional edge cases, like too-big corner radii (as in the screenshot above) and arc start/end values outside of `(0, TWO_PI)` range (this is currently not breaking usage and so should not become breaking usage).

Where out-of-domain parameters to canvas rendering context method is the source of the error, the error provided by p5 will reflect the different function usage:

image

SCREENSHOT: ~Same code running with current p5 (left) and locally (right) demonstrating difference between bad `rect` input of negative radius. The p5 error provides the canvas error, which is different due to refactor, and the new error is not caught correctly. I am not sure where to update the corresponding code.~ Edited to add: sorry, the nice error is just in the p5 editor. In plain browser, both errors are equally uncaught, just the name of the function is different.

*There is an idea in the original issue discussion to use visual automated tests (https://github.com/processing/p5.js/issues/6485#issuecomment-1783671690: "...the idea is to run sample input through p5's current implementation of ellipse and create an image from that. Let's call this the base image. Then, you run the same input through your implementation and create another image. Let's call that the compare image. The test does a pixel-by-pixel comparison of the base image and the compare image. If there's a difference, the test fails.") which I think would be really good, and I would be happy to work on that, but it seems like a separate Issue/PR for workflow enhancement.

### CODE REVIEW

File: src/core/p5.Renderer2D.js [hunk 1]
```diff
@@ -804,14 +745,7 @@ class Renderer2D extends p5.Renderer {
...
 
-      // Draw shape
-      if (!this._clipping) ctx.beginPath();
-      ctx.moveTo(x + tl, y);
-      ctx.arcTo(x + w, y, x + w, y + h, tr);
-      ctx.arcTo(x + w, y + h, x, y + h, br);
-      ctx.arcTo(x, y + h, x, y, bl);
-      ctx.arcTo(x, y, x + w, y, tl);
-      ctx.closePath();
+      ctx.roundRect(x, y, w, h, [tl, tr, br, bl]);
```
Comment (ksen0) [inline src/core/p5.Renderer2D.js:748]: On bad usage (negative radii) this produces a different error than arcTo, so it is not caught correctly. I was not able to find where to make a corresponding update. I looked in the `friendly_errors` directory, particularly in `stacktrace.js`, but was not sure how to proceed.

**Comment by ksen0** [comment]
I added some more inline comments, as it was a bit confusing to think about intended stroke behavior. Here are some of my test cases for comparison:

```
// rotates around, does not fail despite increasing angle values
// no stroke to close the "pie"
arc(200,150,50,80, frameCount/10, PI * 1.5 + frameCount/10); 

// there is a line from center to the right. No visible sliver of the pie
arc(500, 580, 200, 80, 0+0.3, 2*PI+0.3, PIE);

// there is no line, it is visually equivalent to an ellipse or CHORD mode
arc(300, 580, 200, 80, 0, 2*PI-0.001, PIE);
```

And of course paceman example:

```
let biteSize = PI / 16;
let startAngle = biteSize * sin(frameCount * 0.1) + biteSize;
let endAngle = TWO_PI - startAngle + TWO_PI*8;
arc(500, 500, 200, 80, startAngle, endAngle, PIE); 
```

**Comment by limzykenneth** [comment]
Looks good. Thanks!
